cfgov/ask_cfpb/documents.py

```diff
@@ -0,0 +1,508 @@
+        answer_text = self.answer_content_text()
+        full_text = f"{short_answer}\n\n{answer_text}\n\n{self.question}"
+        return full_text
+
+    def __str__(self):
+        if self.answer_base:
+            return f"{self.answer_base.id}: {self.title}"
+        else:
+            return self.title
+
+    @property
+    def clean_search_tags(self):
+        return [
+            tag.strip()
+            for tag in self.search_tags.split(",")
+        ]
+
+    @property
+    def status_string(self):
+        if self.redirect_to_page:
+            if not self.live:
+                return ("redirected but not live")
+            else:
+                return ("redirected")
+        else:
+            return super(AnswerPage, self).status_string
+
+    # Returns an image for the page's meta Open Graph tag
+    @property
+    def meta_image(self):
+        if self.social_sharing_image:
+            return self.social_sharing_image
+
+        if not self.category.exists():
+            return None
+
+        return self.category.first().category_image
+
+    # Overrides the default of page.id for comparing against split testing
+    # clusters. See: core.feature_flags.in_split_testing_cluster
+    @property
+    def split_test_id(self):
+        return self.answer_base.id
+
+
+@registry.register_document
+class AnswerPageDocument(Document):
+
+    autocomplete = fields.TextField(analyzer=label_autocomplete)
+    portal_topics = fields.KeywordField()
+    portal_categories = fields.TextField()
+    text = fields.TextField(attr="text", analyzer=synonym_analyzer)
+    url = fields.TextField()
+    preview = fields.TextField(attr="answer_content_data")
+
+    def get_queryset(self):
+        query_set = super().get_queryset()
+        return query_set.filter(live=True, redirect_to_page=None)
+
+    def prepare_autocomplete(self, instance):
+        return instance.question
+
+    def prepare_portal_categories(self, instance):
+        return [topic.heading for topic in instance.portal_category.all()]
+
+    def prepare_portal_topics(self, instance):
+        return [topic.heading for topic in instance.portal_topic.all()]
+
+    def prepare_search_tags(self, instance):
+        return instance.clean_search_tags
+
+    def prepare_url(self, instance):
+        return instance.url
+
+    class Index:
+        name = 'ask-cfpb'
+        settings = {'number_of_shards': 1,
+                    'number_of_replicas': 0}
+
+    class Django:
+        model = AnswerPage
+
+        fields = [
+            'search_tags',
+            'language',
+        ]
+
+
+def get_suggestion_for_search(search_term):
+    s = AnswerPageDocument.search().suggest(
+        'text_suggestion', search_term, term={'field': 'text'})
+    response = s.execute()
+    try:
+        suggested_term = response.suggest.text_suggestion[0].options[0].text
+        return suggested_term
+    except IndexError:
+        return search_term
+
+
+class AnswerPageSearch:
```

Comment:
I also wonder whether this class could be placed in the `cfgov/search` package and take a model in its constructor. That'd give us a single place where we implement autocomplete, search, suggestions, etc, that would be reusable across all our apps.

@alexm118 @higs4281 do you think that would make sense? I'm basing the idea on a reading of how this works, and I'm not totally confident that reading is accurate.

Reply:
I think we may want to generalize some of this code - like the analyzers into a common file to be referenced across multiple search areas.

Reply:
Moved AnswerPageSearch into cfgov/search package. Code generalization could be done in another PR when other applications start to use ES7.